src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (finalDelimiterNeedsSemicolon)
+            {
+                if (!StatementClosingDelimiterExists(currentNode, ref lastDelimiterSpan))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    nextCommandHandler();
+                    return;
+                }
+            }
+
+            // if you haven't found any enclosures, put semicolon at end of statement
+            if (lastDelimiterSpan == default)
+            {
+                lastDelimiterSpan = currentNode.Span;
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterSpan.End, currentNode.Kind())));
+            nextCommandHandler();
+        }
+
+        private static SyntaxNode GetEnclosingArgumentList(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts)
+        {
+            while (currentNode.Kind() != SyntaxKind.ArgumentList
+                && currentNode.Kind() != SyntaxKind.ArrayRankSpecifier)
+            {
+                if (currentNode.Kind() == SyntaxKind.InterpolatedStringExpression || currentNode.Kind() == SyntaxKind.StringLiteralExpression)
+                {
+                    return null;
+                }
+
+                if (currentNode == null
+                    || syntaxFacts.IsStatement(currentNode)
+                    || currentNode.Kind() == SyntaxKind.VariableDeclaration)
+                {
+                    return null;
+                }
+
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                {
+                    return null;
+                }
+            }
+
+            // now we're in an argument list, so return the enclosing statement
+            while (!syntaxFacts.IsStatement(currentNode) && !(currentNode.Kind() == SyntaxKind.VariableDeclaration))
+            {
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                {
+                    return null;
+                }
+            }
+
+            return currentNode;
+        }
+
+        private static bool IsCaretAtEndOfLine(SnapshotPoint caret)
+        {
+            return caret.Position == caret.GetContainingLine().End;
+        }
+
+        private static SyntaxToken GetToken(SyntaxNode root, int caretPosition, SnapshotPoint caret)
+        {
+            if (IsCaretAtEndOfLine(caret) && caretPosition > 0)
+            {
+                return root.FindToken(caretPosition - 1);
+            }
+            else
+            {
+                return root.FindToken(caretPosition);
+            }
```

Comment:
needs explanation as to why.

Reply:
📝 The tests all pass if the special case here is removed. I need to check with @chborl on why this is included. It might be part of the former implementation of closing delimiter insertion that's been removed for now.

Reply:
If we can't find a case where this is relevant, we should inline `root.FindToken(caretPosition)` to the callers and remove both `GetToken` and `IsCaretAtEndOfLine`.

Reply:
this still seems to apply.